chainer/functions/math/det.py

```diff
@@ -0,0 +1,112 @@
+import numpy
+
+from chainer import cuda
+from chainer import function
+from chainer.functions.array import reshape
+from chainer.functions.math import inv
+from chainer.functions.math import matmul
+from chainer import utils
+from chainer.utils import type_check
+
+
+def _det_gpu(b):
+    # We do a batched LU decomposition on the GPU to compute
+    # and compute the determinant by multiplying the diagonal.
+    # Change the shape of the array to be size=1 minibatch if necessary.
+    # Also copy the matrix as the elments will be modified in-place.
+    a = matmul._as_batch_mat(b).copy()
+    n = a.shape[1]
+    n_matrices = len(a)
+    # Pivot array
+    p = cuda.cupy.zeros((n_matrices, n), dtype='int32')
+    # Output array
+    # These arrays hold information on the execution success
+    # or if the matrix was singular.
+    info1 = cuda.cupy.zeros(n_matrices, dtype=numpy.intp)
+    ap = matmul._mat_ptrs(a)
+    _, lda = matmul._get_ld(a)
+    cuda.cublas.sgetrfBatched(cuda.Device().cublas_handle, n, ap.data.ptr, lda,
+                              p.data.ptr, info1.data.ptr, n_matrices)
+    det = cuda.cupy.prod(a.diagonal(axis1=1, axis2=2), axis=1)
+    # The determinant is equal to the product of the diagonal entries
+    # of `a` where the sign of `a` is flipped depending on whether
+    # the pivot array is equal to its index.
+    rng = cuda.cupy.arange(1, n + 1, dtype='int32')
+    parity = cuda.cupy.sum(p != rng, axis=1) % 2
+    sign = 1. - 2. * parity.astype('float32')
+    success = cuda.cupy.all(info1 == 0)
+    return det * sign, success
+
+
+class BatchDet(function.Function):
+
+    @property
+    def label(self):
+        return 'det'
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+        a_type, = in_types
+        a_type = matmul._convert_type(a_type)
+        type_check.expect(a_type.dtype.kind == 'f')
+        # Only a minibatch of 2D array shapes allowed.
+        type_check.expect(a_type.ndim == 3)
+        # Matrix inversion only allowed for square matrices
+        # so assert the last two dimensions are equal.
+        type_check.expect(a_type.shape[-1] == a_type.shape[-2])
+
+    def forward_cpu(self, x):
+        self.detx = utils.force_array(numpy.linalg.det(x[0]))
+        return self.detx,
+
+    def forward_gpu(self, x):
+        self.detx, success = _det_gpu(x[0])
+        if not success:
+            raise ValueError('Singular Matrix')
```

Comment:
I noticed that we can determine whether we should return 0 without checking `info1`.

According to the [document](http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-getrfbatched), if some element in `info1` is non-zero, corresponding element in the diagonal (`U` in the document and `a` in your implementation) is 0. So the result of `cuda.cupy.prod(a.diagonal(axis1=1, axis2=2), axis=1)` should be 0.